scripts/scrub-user.js

```diff
@@ -0,0 +1,195 @@
+    Total clips contributed: ${stats.clip_total}
+    First contribution: ${stats.first_clip}
+    `);
+};
+
+const deleteClipRecords = (connection, userStats, isLast) => {
+  const idStrings = userStats.ids.map(id => `"${id}"`);
+  connection.query(
+    `DELETE FROM clips WHERE client_id IN (${idStrings})`,
+    (clip_error, clip_results, clip_fields) => {
+      if (clip_error) throw clip_error;
+
+      if (clip_results.affectedRows) {
+        deletionSummary.s3_ids = deletionSummary.s3_ids.concat(userStats.ids);
+
+        if (userStats.in_past_dataset)
+          deletionSummary.dataset_ids = deletionSummary.dataset_ids.concat(
+            userStats.ids
+          );
+
+        console.log(
+          `Records for ${clip_results.affectedRows} clips were deleted.\n`
+        );
+      }
+
+      processIteration(isLast);
+    }
+  );
+};
+
+const deleteUserRecords = (connection, userStats, isLast) => {
+  connection.query(
+    `DELETE FROM user_clients WHERE email = ${userStats.email}`,
+    (delete_error, delete_results, delete_fields) => {
+      if (delete_error) throw delete_error;
+
+      deletionSummary.emails.push(userStats.email);
+
+      if (userStats.clip_total) {
+        readline.question(
+          `\nWould you also like to delete the ${userStats.clip_total} associated clips? \nThis CANNOT be reversed [Y/N] `,
+          answer => {
+            if (answer.toLowerCase() === 'y')
+              deleteClipRecords(connection, userStats, isLast);
+            else processIteration(isLast);
+          }
+        );
+      } else {
+        processIteration(isLast);
+      }
+    }
+  );
+};
+
+const deleteUser = (connection, email, isLast) => {
+  if (safeToDelete) {
+    safeToDelete = false;
+    const clean_email = connection.escape(email);
+
+    connection.query(
+      `
+        SELECT email, u.client_id, COUNT(c.id) clip_count, c.created_at first_clip
+        FROM user_clients u
+        LEFT JOIN clips c ON u.client_id = c.client_id
+        WHERE email = ${clean_email}
+        GROUP BY u.client_id
+        ORDER BY c.created_at ASC;
+      `,
+      (error, results, fields) => {
+        if (error) throw error;
+        const userStats = parseResults(results, clean_email);
+
+        if (userStats.length) {
+          printUserStats(userStats);
+
+          readline.question(
+            `Are you sure you want to delete all data associated with ${clean_email}? \nThis CANNOT be reversed [Y/N] `,
+            answer => {
+              if (answer.toLowerCase() === 'y')
+                deleteUserRecords(connection, userStats, isLast);
+              else processIteration(isLast);
+            }
+          );
+        } else {
+          console.log(
+            `There are no user accounts associated with the email address ${clean_email}.`
+          );
+          processIteration(isLast);
+        }
+      }
+    );
+  } else {
+    setTimeout(() => {
+      deleteUser(connection, email, isLast);
+    }, 200);
```

Comment:
I don't love this piece, but without this the script would print all user stats at once, and accept all prompts at once, which still works but is a real sucky experience prone to error. Super open to hearing a different way of handling this! 

Alternatively, could make this script handle only one email address at a time, but that's pretty inefficient.

Reply:
> Alternatively, could make this script handle only one email address at a time, but that's pretty inefficient.

Is that not what's happening already? `emails.map` (or `emails.forEach`, as recommended below) will process each email sequentially. When the first email runs through `deleteUser`, it will mutate `safeToDelete` and block all other emails until it calls `processIteration` and releases the lock. Soon after, one of the `setTimeout`s will resolve, lock the `safeToDelete` variable, and start the process over again.

To me, it seems like something like:

```
for (const [i, email] of emails.entries()) {
  await deleteUser(connection, email, i === emails.length - 1);
}
```

should clear things up. Note, this will involve converting the functions above to `async`.

Reply:
Sorry I meant "one at a time" as in "you'd have to keep typing `node script email`, instead of being able to do `node script emailA emailB emailC`". I tried implementing this with  async/await, but async/await is [pretty messy](https://zellwk.com/blog/async-await-in-loops/) inside loops. Couldn't get it to work in a `for...of` loop (as in it would print all `readline` prompts at once), it definitely isn't expected to work in a `forEach` loop with a callback, and with asyncs inside `map` I'd need to treat it as a `Promises.all`. Overall I did feel like `setTimeout` was cleaner and easier to follow than turning all callbacks up the chain into `asyncs` and then dealing with the Promise array. Totally open to being persuaded otherwise though.

Reply:
> I meant "one at a time" as in "you'd have to keep typing `node script email`

Ah okay, got it.

> I tried implementing this with async/await, but async/await is pretty messy

[Yeah, that's the classic thing with asynchronous code](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/). Once you do one, you've gotta do 'em all…

> Couldn't get it to work in a for...of loop (as in it would print all readline prompts at once)

It sounds like part of the call chain wasn't converted to `async`, so the `await` was resolving immediately. `for...of` _should_ be able to work smoothly with `async`. However,

> it definitely isn't expected to work in a `forEach` loop with a callback, and with asyncs inside `map` I'd need to treat it as a `Promises.all`.

Yep, it definitely won't work in a `forEach`. But for the same reason, it won't work in `map` either! The separate calls from `map` don't have an easy way to communicate, so there's no (good) way to say "wait for the previous iteration". `Promise.all` helps if all the calls can run at once, but the next step needs to wait until everything has resolved. Here, we need to schedule one after another – hence `for...of`.

> Overall I did feel like `setTimeout` was cleaner and easier to follow than turning all callbacks up the chain into `asyncs`

To each their own! If you want to pair on this sometime this week, it could be a fun chance for us to work together 😊 this seems to work as-is, though, so I don't wanna block you any longer.

Thanks for your patience and for the discussion! Clicking the big green button in 3... 2... 1...